### DISCUSSION TITLE
Queries about using PyBricks in the WRO open and international competitions

### DISCUSSION THREAD

**Discussion by MohammedEsmaeel** [pybricks/support#1712] [Q&A]
Hello All,

1-What **approach** should a team take when using PyBricks at the WRO competition?

2-How should a team **inform** the judges that they will be **using Bluetooth**?

3-What should a team do if the **judges are unaware that Bluetooth usage is allowed**?

4-Have there been any **issues** between teams and judges regarding the use of **Bluetooth**?

5-Has PyBricks **communicated directly** with WRO about Bluetooth being the only way to use your framework?

6-Can you provide **statistics** on the use of PyBricks in local and international WRO robot tournaments?

And thanks for you all.

**Comment by laurensvalk** [comment]
> What approach should a team take when using PyBricks at the WRO competition?

The same as with any other app --- work hard and play well. You could think of Pybricks as a useful tool that is specifically designed to help you do that :slightly_smiling_face: 

You can read more about the specifics and some insights from teams in our new [learning guide](https://pybricks.com/learn/).

> Has PyBricks communicated directly with WRO about Bluetooth being the only way to use your framework?

I have personally spoken with the global WRO organization. They like Pybricks, and don't seem to have issues with Bluetooth when used appropriately. That said, local organizers sometimes take a more strict view, so always check the local rules or check with the organizers.

> How should a team inform the judges that they will be using Bluetooth? (...) What should a team do if the judges are unaware that Bluetooth usage is allowed? (...) Have there been any issues between teams and judges regarding the use of Bluetooth?

Others like @afarago and @ggramlich have much more experience with judging and working with judges, so they may be able to provide more insight.

We are also working on a new feature that makes it easy to turn off Bluetooth altogether.

> Can you provide statistics on the use of PyBricks in local and international WRO robot tournaments?

It's difficult to get exact numbers since we don't hear from every team, but WRO teams have done really well with Pybricks in many finals this year, including @ggramlich. In some categories in some countries, all the top teams with Spike used Pybricks.

The same is true for FLL. For example, out of 1900 teams in the UK, the top two teams both used Pybricks ([source](https://x.com/E_Dave42/status/1789744851988172862)).

**Comment by afarago** [comment]
Hey @MohammedEsmaeel , glad to see more coaches and teams joining the discussion!
I can mostly just confirm and repeat what @laurensvalk said, but let's get into the details.

> 1-What **approach** should a team take when using PyBricks at the WRO competition?

Just do it and experiment!
Take extra tasks and be brave to try things many times and do not skip to the next step until the current one is not working in a rock solid way.
Laurens is starting up a tutorial also I am planning to create a brief and hands on WRO+FLL targeted starter info kit by August.

> 2-How should a team **inform** the judges that they will be **using Bluetooth**?

Actually as a WRO judge / head judge in some of the competition it is obvious that the current rule of "turn off BT" is impossible even by the SPIKE hubs. BT is always on even on the stock firmware, no matter what the kids do, it is only not connectable by public devices and can be disconnected from any computer. This is what we ask for.
Generally we insist on programming by cable if possible. If a team comes with e.g. a tablet this is obviously impossible even by the stock firmware and the current version of Pybricks as well.
Best practice: coach writes an email beforehand to the organizers, kids tell the judge on the spot at the entrance during the entrance check.

> 3-What should a team do if the **judges are unaware that Bluetooth usage is allowed**?

Same as with all competitions. Have the rulebook printed with them and point to the right place defending their position - kindly and with respect.
If they are right, they should know it and should be able to enter -even slightly more difficult- conversation if needed.
Some of us will also be in Izmir so hopefully can channel more info back in November.

> 4-Have there been any **issues** between teams and judges regarding the use of **Bluetooth**?

None I am aware of.
Us, judges are always nice persons seeking ways to cheer up kids and up to the last resort and possibility give the most point allowed by the rules. Volunteering and spending our free time is best rewarded when kids can do their best at the table!

> 5-Has PyBricks **communicated directly** with WRO about Bluetooth being the only way to use your framework?

See the info from Laurens, but this is not necessary as this year it is not Laurens or David competing. :D

> 6-Can you provide **statistics** on the use of PyBricks in local and international WRO robot tournaments?

I am seeing a few teams both in FLL and WRO with alternative firmwares.
This even varies from a ev3dev to a Russian C based EV3 (cannot recall the name) environment or ev3basic, but also you can use text based own interpreters or pybricks or any firmware you prefer.
Some teams are successful, some are less, still pybricks provides a huge confidence and right learning curve.
In my experience (2 years with Pybricks, 10 years with FLL/WRO) this gives a way better time investment-success ratio than the stock firmware.

Hope this helps, feel free to ask if anything is unclear!

**Reply by ogzhnkse** [reply]
Hi,

We’ve decided to use Pybricks in Izmir and also for the FLL Supermerged season. However, we’re concerned about the rule requiring Bluetooth to be turned off. This means we can’t send code to the robot via USB cable, and using a tablet would pose the same issue. If you’re waiting for an answer from the National Organizer in Turkey, don’t expect a positive response, as I’m one hundred percent sure they don’t understand what we’re talking about. Please don’t take this as disrespect; there really isn’t anyone involved in the organization who understands these matters.

**Comment by afarago** [comment]
I will not be able to responsibly answer your question now... but if we are lucky we might even meet at Izmir as I will have the honor to be serving as a table judge there. Which team are you coaching?

Out of curiosity - have you asked this question in email? If not I would strongly advise to do so asap. I think you can trust the WRO head panel's robotics experience.

What I can do is to also address this and ask our head judge parallel.

By the way - I guess you are aware that the new firmware(beta?) has the option to switch off Bluetooth for the competition round so the kids need a permit for programming via BT instead of cable; right?

**Comment by ggramlich** [comment]
@ogzhnkse 
We will participate in Izmir in the Elementary category. We are using the Beta Firmware which allows to switch off Bluetooth and has a red / blue indicator as shown in https://github.com/pybricks/support/issues/1615#issuecomment-2147670765

As far as I have followed it, this feature is not in the stable firmware yet.
See https://pybricks.com/learn/getting-started/install-pybricks/#trying-the-nightly-version-optional for the recent builds of the firmware.

I will make our German organizers aware of this github issue and ask them to pose the question to the international organizers. Especially for the teams with the original firmware who can not switch off bluetooth.
As @afarago mentioned: Please also pose the question yourself.

Also make sure that on thursday the 28th in the "Free Practice Time" your team already communicates with their table judges so that they can understand which options are supported by (your version of) Pybricks. On that day it might be much easier to get the coach involved when questions arise than on the acutal Competition days.

@afarago It would be nice to meet you. Do you already know which category you are judging?

**Reply by ogzhnkse** [reply]
@ggramlich 
Thank you very much for the information. Actually, since the finals in Germany, I have been constantly warning the head judges to make an arrangement for Spike and Inventor, because these robots also have charging problems, I think the second hub should be allowed.
While the children are working they need to tell the judges that they need to send codes via BT but that they need to turn BT off during the competition.
I am participating as a coach  from Turkey, so I know the national organizers, that's why I wrote that.
I also have teams in four different categories: elementary, junior, future engineer and future innovators.
We are looking forward to your coming to Turkey, by the way, our school is in Izmir, so you are coming to our house, and if there is anything you want to learn in advance and I can help you, please write to me.

**Reply by enFranz** [reply]
This is a very interesting and important discussion. I understand, it will become difficult to work with PyBricks and there is the risk, that PyBricks will be not allowed in Izmir. For my team from Austria, this would kill any chance of a successful participation.
One comment said, it would make sense to get permission upfront an other opinion is, it would not be possible.
Is the Head of Judge already put into duty? Does someone know how to contact Head of Judge?
Or would it be contra productive to ask for approval?

From my point of view the rules (German Version) say, that Bluetooth instead of cable is allowed, if the "device" does not support cable connection. What is needed to prove, that cable is not supported?

I appreciate any comment to that - KR

**Reply by afarago** [reply]
Hey @enFranz @ggramlich @ogzhnkse Glad to see so many coaches around!
Should we plan for a cool pybricks-fans photo in Izmir for Laurens and David? :D

The argument might nigh be enough as actually the device itself supports cable connection, just the custom firmware does not.
I know that before the competition new firmware is not a good idea, but if I were prepping a team now, I would strongly consider at least the beta firmware where the BT can be turned off, so you at least avoid the risk that during the quarantine or competition phase an external source accidentally overwrites the program you the team's robot. Currently in pybricks there is no PIN or no-discovery protection yet.

The head judges are in already named, and we will have a good level of professionalism there. However I am not entitled to give out the contact.
Please go through the official organizer email channel and pose this question upfront.

Hope this helps, happy prepping, see you in Izmir!

**Reply by enFranz** [reply]
Hi @afarago, thanks for your prompt reply.
Yes it is a risk to change firmware version close to the final, but I understand your good argument and I will give it a try - at least as backup.
I hope, we find and meet in Izmir - good for sharing experience and most likely for pybricks fan picture.

I am looking forward, seeing you in Izmir

**Reply by ogzhnkse** [reply]
Hi @afarago,
I agree the @enFranz beta firmware will be risky but we have to I think 👍 I live in İzmir as I said if you need anything please feel free to write me. Fan picture good idea 😁 I hope in these picture our face will have with big smile (after the coach meeting) By the way I give an info to Turkish head judges and organization partners about this stiuation.

**Reply by ggramlich** [reply]
@ogzhnkse I received this answer from the German contact:

> Mach Dir bez. PyBricks keine Sorgen. Die WRO weiß über PyBricks Bescheid – das ist ja auch kein Nischenprodukt mehr – und ich denke nicht, dass es da Probleme geben wird. 

> Don't worry about PyBricks. The WRO knows about PyBricks - it's no longer a niche product - and I don't think there will be any problems with it.

Thanks for the invitation.

**Reply by enFranz** [reply]
@ggramlich - thanks for your respond 😁
See you, KR

**Reply by ogzhnkse** [reply]
@ggramlich Thank you for letting us know. 😊🙏

**Comment by enFranz** [comment]
Hi, I have installed the Beta-Version, "the nightly version" (https://pybricks.com/learn/getting-started/install-pybricks/). In terms of the Bluetooth functions, I like the implementation:
After power on - no Bluetooth connection ... it is necessary to press the Bluetooth button for pairing
At the PC/Laptop, I decide for connect (pairing) or disconnect (Bluetooth button is blinking)

Bluetooth button is "off" means:
- for the judge, there is no option to modify the software
- it protects the respective team against unexpected changes of the loaded software.

I, we, did no specific tests with our robot application so far. I hope, there are no drawbacks at this side.
If anyone has experience with this version (v3.6.0b2), I appreciate any feedback.
KR

**Reply by ogzhnkse** [reply]
In the national tournament in Turkey, we experienced an issue where, during test runs, others connected from different computers and unfortunately deleted the program inside. Is there a way to prevent this? By the way we used prime firmware not pybricks but the same issue might happen to us.

**Reply by afarago** [reply]
Actually with Prime firmware I cannot imagine how one could accomplish this.
As far as I recall Prime firmware asks for a BT button push before a new PC connect, so without that an unknown source has no chance to connect your brick.

By the way - advice to the ones coming to the WRO finals: make sure to add a unique name to your bot, that others will not accidentally connect to.

**Comment by afarago** [comment]
Hope all your teams were successful this year as well and might meet some of you in Singapore WRO 2025!
Anyone interested to meet and greet; teams exchanging best practices; just let's connect!

Additionally: in case you are in for some experiment - would be happy to get some feedback to a vscode extension supporting pybricks mostly for advenced/competition users and teams. https://github.com/orgs/pybricks/discussions/2389
